marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/reader/HubInputPartitionReader.java

```diff
@@ -88,10 +88,21 @@ public void close() {
         logger.debug("Closing");
     }
 
-    private ObjectNode buildEndpointConstants(JsonNode initializationResponse, int partitionNumber) {
+    private ObjectNode buildEndpointConstants(Map<String, String> options, JsonNode initializationResponse, int partitionNumber) {
         ObjectNode endpointConstants = objectMapper.createObjectNode();
         endpointConstants.set("initializationResponse", initializationResponse);
         endpointConstants.put("partitionNumber", partitionNumber);
+        String optimizationlevel = options.get("optimizationlevel");
```

Comment:
Given that we have a couple levels of nesting going on here, that's a good indicator that we could use a new private method for readability here - e.g. 

```
addOptimizationLevel(options, endpointConstants);
```

That method will then do the error handling and validation of the value that the user provides.